### PR TITLE
Fix global search; closes #7514

### LINKED ISSUE #7514: [RC feedback] Global Search doesn't show results - Class 'AllAssets' not found
**Describe the bug**

Global Search doesn't show any results and there are errors in the browser as well as in the logs. Looks like it's related to #7286 #7294 but I use the version 9.5 RC2 where it should be fixed. I do see the changes of the mentioned bugfixes in the search.class.php.
I have the 'Objects management' plugin installed but the error appears also if it's deactivated.
Also, I changed the columns prior to the update from 9.4.6 to 9.5 RC2 so that it should show: Name, Status, Comments, Network - IP, Type of entry. In 9.4.6 it worked fine.

**To reproduce**

1. Open Assets > Global
2. The page doesn't load.

**Logs**

php-errors.log shows:
```
[2020-07-01 11:24:19] glpiphplog.CRITICAL:   *** Uncaught Exception Error: Class 'AllAssets' not found in /volume1/web/glpitest/inc/search.class.php at line 561
  Backtrace :
  inc/search.class.php:97                            Search::constructSQL()
  inc/search.class.php:80                            Search::showList()
  front/allassets.php:45                             Search::show()
 
```

**Screenshots**

image

### PR DESCRIPTION
**Pull request by trasher**
| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #7514 

since 2ee3984be7b7374f7201a7a13a2273b953e7a408, calling `getTableForItem` will work in every case; `::getTable` requires an existing class.
Current fix is a partial revert of that commit.

Note: we would need unit tests on 'AllAsset' :-/

### CODE REVIEW

File: inc/search.class.php [hunk 1]
```diff
@@ -558,7 +558,7 @@ static function constructSQL(array &$data) {
       $blacklist_tables = [];
       if (isset($CFG_GLPI['union_search_type'][$data['itemtype']])) {
          $itemtable          = $CFG_GLPI['union_search_type'][$data['itemtype']];
-         $blacklist_tables[] = $data['itemtype']::getTable();
+         $blacklist_tables[] = getTableForItemType($data['itemtype']);
```
Comment (trasher) [inline inc/search.class.php:561]: or maybe:
```suggestion
         $blacklist_tables[] = (class_exists($data['itemtype']) ?  $data['itemtype']::getTable() : getTableForItemType($data['itemtype']));
```

**Comment by trasher** [comment]
OK, I do not know how to address this one without making a lot of local changes in Search once again...

File: inc/search.class.php [hunk 2]
```diff
@@ -556,11 +549,12 @@ static function constructSQL(array &$data) {
       $searchopt        = &self::getOptions($data['itemtype']);
 
       $blacklist_tables = [];
+      $orig_table = ($data['itemtype'] == 'AllAssets' ? getTableForItemType($data['itemtype']) : $data['itemtype']::getTable());
```
Comment (cedric-anne) [inline inc/search.class.php:552]: You may also break reservation search here, or anything that uses `$CFG_GLPI['union_search_type']`.
```suggestion
      $orig_table = (!is_a($data['itemtype'], CommonDBTM::class, true) ? getTableForItemType($data['itemtype']) : $data['itemtype']::getTable());
```

File: inc/search.class.php [hunk 3]
```diff
@@ -3213,7 +3207,8 @@ static function addOrderBy($itemtype, $ID, $order) {
 
       $is_fkey_composite_on_self = getTableNameForForeignKeyField($searchopt[$ID]["linkfield"]) == $table
          && $searchopt[$ID]["linkfield"] != getForeignKeyFieldForTable($table);
-      if (($is_fkey_composite_on_self || $table != $itemtype::getTable())
+      $orig_table = ($itemtype == 'AllAssets' ? getTableForItemType($itemtype) : $itemtype::getTable());
```
Comment (cedric-anne) [inline inc/search.class.php:3210]:
```suggestion
      $orig_table = (!is_a($itemtype, CommonDBTM::class, true) ? getTableForItemType($itemtype) : $itemtype::getTable());
```

File: inc/search.class.php [hunk 4]
```diff
@@ -3363,7 +3358,7 @@ static function addDefaultToView($itemtype, $params) {
    static function addDefaultSelect($itemtype) {
       global $DB;
 
-      $itemtable = $itemtype::getTable();
+      $itemtable = ($itemtype == 'AllAssets' ? getTableForItemType($itemtype) : $itemtype::getTable());
```
Comment (cedric-anne) [inline inc/search.class.php:3361]:
```suggestion
      $itemtable = (!is_a($itemtype, CommonDBTM::class, true) ? getTableForItemType($itemtype) : $itemtype::getTable());
```

File: inc/search.class.php [hunk 5]
```diff
@@ -3430,7 +3425,9 @@ static function addSelect($itemtype, $ID, $meta = 0, $meta_type = 0) {
 
       $is_fkey_composite_on_self = getTableNameForForeignKeyField($searchopt[$ID]["linkfield"]) == $table
          && $searchopt[$ID]["linkfield"] != getForeignKeyFieldForTable($table);
-      if (((($is_fkey_composite_on_self || $table != $itemtype::getTable())
+
+      $orig_table = ($itemtype == 'AllAssets' ? getTableForItemType($itemtype) : $itemtype::getTable());
```
Comment (cedric-anne) [inline inc/search.class.php:3429]:
```suggestion
      $orig_table = (!is_a($itemtype, CommonDBTM::class, true) ? getTableForItemType($itemtype) : $itemtype::getTable());
```

File: inc/search.class.php [hunk 6]
```diff
@@ -4092,8 +4089,9 @@ static function addWhere($link, $nott, $itemtype, $ID, $searchtype, $val, $meta
       $addtable  = '';
       $is_fkey_composite_on_self = getTableNameForForeignKeyField($searchopt[$ID]["linkfield"]) == $table
          && $searchopt[$ID]["linkfield"] != getForeignKeyFieldForTable($table);
+      $orig_table = ($itemtype == 'AllAssets' ? getTableForItemType($itemtype) : $itemtype::getTable());
```
Comment (cedric-anne) [inline inc/search.class.php:4092]:
```suggestion
      $orig_table = (!is_a($itemtype, CommonDBTM::class, true) ? getTableForItemType($itemtype) : $itemtype::getTable());
```

File: inc/search.class.php (hunk 2 above)
Comment (trasher) [reply inc/search.class.php:552]: OK, this need to be factorized :)

**Pull request merged**